iml-api/src/graphql.rs

```diff
@@ -478,6 +559,92 @@ impl QueryRoot {
 
         Ok(xs)
     }
+
+    #[graphql(arguments(
+        limit(description = "optional paging limit, defaults to all rows"),
+        offset(description = "Offset into items, defaults to 0"),
+        dir(description = "Sort direction, defaults to asc"),
+        message(description = "Pattern to search for in message"),
+        fqdn(description = "Pattern to search for in FQDN"),
+        tag(description = "Pattern to search for in tag"),
+        start_datetime(description = "Start of the time period of logs"),
+        end_datetime(description = "End of the time period of logs"),
+        message_class(description = "Array of log message classes"),
+        severity(description = "Upper bound of log severity"),
+    ))]
+    async fn logs(
+        context: &Context,
+        limit: Option<i32>,
+        offset: Option<i32>,
+        dir: Option<SortDir>,
+        message: Option<String>,
+        fqdn: Option<String>,
+        tag: Option<String>,
+        start_datetime: Option<String>,
+        end_datetime: Option<String>,
+        message_class: Option<Vec<MessageClass>>,
+        severity: Option<LogSeverity>,
+    ) -> juniper::FieldResult<Vec<LogMessage>> {
+        let dir = dir.unwrap_or_default();
+
+        let start_datetime = start_datetime
+            .map(|s| s.parse::<chrono::DateTime<Utc>>())
+            .map_or(Ok(None), |r| r.map(Some))?;
+
+        let end_datetime = end_datetime
+            .map(|s| s.parse::<chrono::DateTime<Utc>>())
+            .map_or(Ok(None), |r| r.map(Some))?;
+
+        let message_class: Vec<_> = if let Some(mc) = message_class {
+            if !mc.is_empty() {
+                mc
+            } else {
+                vec![MessageClass::Normal]
+            }
+        } else {
+            vec![MessageClass::Normal]
+        }
```

Comment:
```suggestion
        let message_class = message_class.filter(|v| !v.is_empty()).unwrap_or_else(|| vec![MessageClass::Normal])
```
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1e6cf53f488c093152a22019aeb84281

Reply:
Done